### DISCUSSION TITLE
Usage of docker containers in Kitware/trame/docker

### DISCUSSION THREAD

**Discussion by earlsioson** [Kitware/trame#50] [Q&A] [answered]
Greetings,

I'm a new user of Paraview/VTK/trame and I'm trying to run one of the trame docker images published on docker hub. I'm running it on a google cloud compute engine vm with an NVIDIA Tesla T4 gpu. I've configured the machine with the nvidia-container-toolkit.

Running the command...

`sudo docker run --gpus all --rm nvidia/cuda:9.0-base nvidia-smi` 

yields...

<img width="884" alt="image" src="https://user-images.githubusercontent.com/13466524/161967352-37979fb6-ba7b-4143-bfd0-d3c3d1edeed0.png">

From looking at the github workflow action, I'm guessing I need to use the kitware/trame:glvnd tag per https://github.com/Kitware/trame/blob/master/.github/workflows/docker.yml#L84

When running it I get this error...

`ERROR: The deploy directory must be mounted into the container at /deploy`

I've tried to trace back what is required by reading through the docker scripts directory and see I need to have a server and setup directory in a mounted deploy directory.

My questions are:

- Can someone point me to an example of what should be in the setup and server directories?
- Am I using the correct image tag, kitware/trame:glvnd, for my machine?

We would like to eventually deploy an app to our Google Cloud Kuberenetes Engine cluster but want to use the Google Cloud Compute Engine virtual machine as a tool for learning how to use Paraview/VTK/trame.

**Comment by jourdain** [comment, marked as answer]
Hi @earlsioson ,

The base image that you are using is indeed a good one to enable remote rendering with ParaView/EGL. But the "trame" images are generic in the sense they only provide the infrastructure to deploy "trame app(s)" into the cloud (Python3 + web routing). They don't even have trame in them. Basically they rely on a mounted directory (/deploy) to provide the actual "trame app(s)" as a virtual environment.  

To get started in how to build/deploy a trame app using such image, you will need to create a "trame application" using our [cookiecutter](https://github.com/Kitware/trame-cookiecutter). In doing so, you will get a `/bundles/docker/*` directory with an example of a `./setup/` directory that get used to build the `./server -> /deploy` expected directory. You can refer to the `/bundles/docker/scripts/*` to see how to build and run using that side directory or just create an image on top that to bundle such directory inside a new image. You can also look at the one we did for [the new visualizer](https://github.com/Kitware/paraview-visualizer/tree/master/docker) too.

We are definitely lacking doc on that docker topic, but we are also trying to make it straight forward using the cookiecutter.

Anyhow, I hope that help you getting started.

**A side note:**
 `pywebvue 2.9.15` and `2.10.0` are currently breaking the local rendering used within the cookiecutter. If you want to try with the default example generated by the cookiecutter you should force `pywebvue==2.9.14` for now.

**Reply by earlsioson** [reply]
Thank you for the help @jourdain. I appreciate it. I did try the new visualizer you linked me to on macOS prior to trying it in the vm. I followed the instructions in the https://github.com/Kitware/paraview-visualizer/tree/master/docker README and get the following...

<img width="893" alt="image" src="https://user-images.githubusercontent.com/13466524/162063712-8327d289-be73-4a54-a569-136782bb9c55.png">

In the browser I navigate to http://localhost:8080/ and get this.
<img width="405" alt="image" src="https://user-images.githubusercontent.com/13466524/162063909-6e5bc9fc-97a6-47e4-9e20-1d4960983d47.png">

Is there a step that I am missing? I did run the build_image.sh script successfully prior to running the run_image.sh. Also, I did have to remove the `--gpus all` argument in run_image.sh so that it would run on macOS, not sure if that is why I'm getting the 404 from apache.

**Reply by jourdain** [reply]
On mac the rendering will not work no matter what unless you replace the EGL paraview by an OSMesa one assuming your mac is an x86. Hope that make sense.

**Reply by jourdain** [reply]
For the gke, I would think it should have worked. Normally apache is supposed to serve `/deploy/server/www`. Does that directory exist?

**Reply by earlsioson** [reply]
In the ubuntu vm, the server dir was created but the www directory is empty...

<img width="1884" alt="image" src="https://user-images.githubusercontent.com/13466524/162069952-d0ff4218-dea5-4796-9a29-7fcbba59857f.png">

**Reply by jourdain** [reply]
I think the build of the server directory may have failed. Can you try again and look for error? (You will need to delete that server directory to force the system to rebuild it.)

**Reply by earlsioson** [reply]
You're right the build did fail. It can't find paraview...

<img width="1884" alt="image" src="https://user-images.githubusercontent.com/13466524/162077391-dcaa99e8-949d-4a53-b01c-32755fc8ccd3.png">

I do have the paraview binaries in a /opt/paraview dir and the bin /opt/paraview/bin in my path and have a PV_VENV defined. The trame/Paraview/SimpleCone/RemoteRendering.py example does work with pvpython. Does the Docker container use the host paraview environment somehow?

**Reply by jourdain** [reply]
I must have broke it in some of my latest changes. Basically the system ask `trame` to fill the `www` directory for each app exposed and because the paraview import is not properly isolated, that step fails. To make it work, I had to guard all the paraview import so the `--www ...` arg usage could actually work. I need to find a better way of handling that. To some extent, this will be easily solved once a coming PR will be merge which will allow to point to a ParaView to extend the current Python environment. At that point, the build step will need a ParaView installed somewhere with the proper env variable set (TRAME_PARAVIEW).

**Reply by earlsioson** [reply]
So I tried using the cookie cutter approach instead and made a little more progress I think. I followed the instruction in the generated trame-app/bundles/docker/README.md and ran the build_server.sh then run_server.sh...
<img width="1904" alt="image" src="https://user-images.githubusercontent.com/13466524/162091056-644827b7-fd1b-4729-8abc-7143a0b3c570.png">

<img width="1781" alt="image" src="https://user-images.githubusercontent.com/13466524/162091086-c8ecff9c-791d-4d0b-83f8-8a76e55dbd3b.png">

I'm assuming that the
`Uncaught TypeError: t.renderable.get_tmAtlas is not a function` error is because of the pywebvue 2.9.15 and 2.10.0 breaking changes you mentioned previously. I'm not sure though how to set the pywebvue==2.9.14 in the cookie cutter.

**Reply by jourdain** [reply]
Indeed this is the error I was mentioning before. So within the generated project edit your [`setup.cfg`](https://github.com/Kitware/trame-cookiecutter/blob/master/%7B%7Bcookiecutter.package_name%7D%7D/setup.cfg#L32-L41) to add `pywebvue==2.9.14`.
Then delete the `/bundles/docker/server` directory and build it again.

**Reply by earlsioson** [reply]
Thank you @jourdain! That fixed the error and I now have a cone rendered in the browser using the trame docker image via the cookiecutter :)

<img width="1293" alt="image" src="https://user-images.githubusercontent.com/13466524/162180232-4861978c-5701-4799-929f-034f69123fa0.png">

One other issue I'm having is when connecting to the service through a front proxy. We are using Envoy setup as a front proxy load balancer performing ssl termination to our backend services. The browser client is trying to setup a websocket connection using the ws protocol but we need to change it so that it tries using the wss protocol instead. Where can I make this change?

<img width="1293" alt="image" src="https://user-images.githubusercontent.com/13466524/162179433-7adda9bd-e015-4164-8a34-6a495d0826f3.png">

**Reply by earlsioson** [reply]
@jourdain I managed to get it to work by changing it in the generated launcher.json. I think I'm good for now. Really appreciate the help.

**Reply by jourdain** [reply]
FYI `pywebvue>=1.5.1` fix the vtk.js rendering issue that the cookiecutter was running into by default...

**Reply by earlsioson** [reply]
@jourdain did you mean pywebvue>=2.10.1 https://github.com/Kitware/py-web-vue/releases/tag/v2.10.1

**Reply by jourdain** [reply]
oups yes, I was reading the version of `vue-vtk-js` which was the underneath source of that issue.

**Reply by earlsioson** [reply]
Hello @jourdain,

On our remote VM we have a trame app that works fine using vtkLocalView but when we try an app that uses vtkRemoteView, we receive the following...
`
Thread 1 "python" received signal SIGSEGV, Segmentation fault.
0x00007fffe86327bf in XSync () from /lib/x86_64-linux-gnu/libX11.so.6
(gdb)
`
On our desktops both of those api calls work but on our remote VM, vtkRemoteView causes that Segmentation fault. Can you help us understand that error?

**Reply by jourdain** [reply]
That is expected unless the rendering backend of the server is relying on EGL or OSMesa. For EGL, you need an nvidia GPU with the proper gpu flag for docker and the nvidia drivers on the host.

Since we don't have a VTK with EGL, you will either have to build it yourself, or use our paraview-egl binary to run a vtk based app.

**Reply by jourdain** [reply]
Basically the server does not have X and the default build is using X to create the GL context. That works fine on a "desktop" machine but not on an headless machine.

**Reply by earlsioson** [reply]
Thanks again @jourdain. Our VM has an nvidia tesla T4 and does have the nvidia drivers installed. We also have the paraview-egl binaries installed in an /opt/paraview directory and the pvpython trame/examples/ParaView/ContourGeometry/RemoteRendering.py example does work on the host machine. Will we have to copy those binaries into the trame:glvnd image when using the generated trame-app from cookiecutter?

**Reply by jourdain** [reply]
Just look at the visualizer docker again. That would be something like that. I haven't tested yet the latest changes I made in visualizer just yet. [will work](https://github.com/Kitware/paraview-visualizer/blob/abb532848323fa21c0574e9c8df2d98d0b7fc2e5/docker/setup/apps.yml) / [might work](https://github.com/Kitware/paraview-visualizer/blob/master/docker/setup/apps.yml)

**Reply by jourdain** [reply]
Basically use `pvpython` instead of `python` in the launcher and `import paraview.web.venv` in your code.

**Reply by earlsioson** [reply]
@jourdain that helped. I think we can use paraview-visualizer as a starting point.  In the visualizer repo I created a PR that merges the latest scripts from cookiecutter output and tweaked for paraview-visualizer in case you wanted to merge into master. I tested it on our vm with the nvidia tesla t4 gpu. https://github.com/Kitware/paraview-visualizer/pull/3
